portal/flatpak-portal.c

```diff
@@ -700,6 +741,466 @@ authorize_method_handler (GDBusInterfaceSkeleton *interface,
+
+  G_UNLOCK (update_monitors);
+}
+
+/* Runs on main thread */
+static gboolean
+check_all_for_updates_cb (void *data)
+{
+  g_autoptr(GTask) task = g_task_new (NULL, NULL, NULL, NULL);
+
+  g_debug ("Checking all update monitors");
+
+  G_LOCK (update_monitors);
+  update_monitors_timeout = 0;
+  update_monitors_timeout_running_thread = TRUE;
+  G_UNLOCK (update_monitors);
+
+  g_task_run_in_thread (task, check_all_for_updates_in_thread_func);
+
+  return G_SOURCE_REMOVE; /* This will be re-added by the thread when done */
+}
+
+/* Runs in worker thread */
+static gboolean
+handle_create_update_monitor (PortalFlatpak *object,
+                              GDBusMethodInvocation *invocation,
+                              GVariant *options)
+{
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  g_autoptr(PortalFlatpakUpdateMonitorSkeleton) monitor = NULL;
+  const char *sender;
+  g_autofree char *sender_escaped = NULL;
+  g_autofree char *obj_path = NULL;
+  g_autofree char *token = NULL;
+  g_autoptr(GError) error = NULL;
+  int i;
+
+  if (!g_variant_lookup (options, "token", "s", &token))
+    token = g_strdup_printf ("%d", g_random_int_range (0, 1000));
+
+  sender = g_dbus_method_invocation_get_sender (invocation);
+  g_debug ("handle CreateUpdateMonitor from %s", sender);
+
+  sender_escaped = g_strdup (sender + 1);
+  for (i = 0; sender_escaped[i]; i++)
+    {
+      if (sender_escaped[i] == '.')
+        sender_escaped[i] = '_';
+    }
+
+  obj_path = g_strdup_printf ("/org/freedesktop/portal/Flatpak/UpdateMonitor/%s/%s",
+                              sender_escaped,
+                              token);
+
+  monitor = (PortalFlatpakUpdateMonitorSkeleton *) create_update_monitor (invocation, obj_path, &error);
+  if (monitor == NULL)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  g_signal_connect (monitor, "handle-close", G_CALLBACK (handle_close), NULL);
+  g_signal_connect (monitor, "g-authorize-method", G_CALLBACK (authorize_method_handler), NULL);
+
+  if (!g_dbus_interface_skeleton_export (G_DBUS_INTERFACE_SKELETON (monitor),
+                                         connection,
+                                         obj_path,
+                                         &error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  register_update_monitor ((PortalFlatpakUpdateMonitor*)monitor, obj_path);
+
+  portal_flatpak_complete_create_update_monitor (portal, invocation, obj_path);
+
+  return TRUE;
+}
+
+/* Runs in worker thread */
+static gboolean
+handle_close (PortalFlatpakUpdateMonitor *monitor,
+              GDBusMethodInvocation *invocation)
+{
+  update_monitor_close (monitor);
+
+  g_debug ("handle UpdateMonitor.Close");
+
+  portal_flatpak_update_monitor_complete_close (monitor, invocation);
```

Comment:
So in the case that the monitor is 'running', we end up returning before it is actually closed. I guess it doesn't matter

Reply:
We don't want to close it while there is random code in some other thread touching it.
However, we probably want to avoid sending any signals once its closed, so that its "effectively" closed.